src/Symfony/Component/Messenger/Command/AbstractFailedMessagesCommand.php

```diff
@@ -30,18 +31,27 @@ abstract class AbstractFailedMessagesCommand extends Command
 {
     private $receiverName;
     private $receiver;
+    /**
+     * @var ServiceLocator
+     */
+    private $failureTransports;
 
-    public function __construct(string $receiverName, ReceiverInterface $receiver)
+    public function __construct(?string $receiverName, ?ReceiverInterface $receiver, ServiceLocator $failureTransports)
```

Comment:
The purpose of `$receiverName` now is to be the `$globalFailureReceiverName`. I think we should rename the variable for clarity. The idea would be that, *if* this is set AND no `--failed-transport=` option is passed, then we will use this automatically. If this is NOT set, then we could ask the user interactively which failure receiver they want to use.

For the 2nd and 3rd arguments, I think they should be "merged". What I mean is, the "default failure receiver" (if there is one) will be inside of the `ServiceLocator $failureTransports` (if it's not already, it should be). This means that we only need 1 argument here, not 2.

We also need to worry about backwards-compatibility and triggering deprecation notices. So I would:

A) Only have a 2nd argument (not a 2nd and third)
B) Remove the type-hint from the argument (so that it allows both a `ReceiverInterface` or a `ServiceLocator`
C) Inside the constructor, check to see if the argument is `!instanceof ServiceLocator`. In that case, trigger a deprecation warning that says that they SHOULD now pass a locator of failure transports. Then write some code to make it work both ways - one simple way might be to create your own `ServiceLocator` object if they're using the deprecated code path - maybe something like:

```
$this->failureTransports = new ServiceLocator([$this->receiverName => function() { return $failureTransports; }
```

I'm not sure if that code to instantiate the ServiceLocator is 100% correct. But the idea is to create a mini-container with the ONE receiver to make the deprecated code path work. Then the rest of the code can work the same.

Reply:
Already talked with @weaverryan about the solution